## 4.9 Questions Part 1

## Importing Libraries

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

## Importing Dataframes

In [2]:
#Pathway to the Dataframes
path = r'C:\Users\Dallas\Desktop\Data Analyst - Projects\Data Immersion\Data Immersion A4\04-2023 Instacart Basket Analysis'

In [3]:
#Importing customers.csv dataframe
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
#Importing orders_products_merged_grouping.pkl dataframe
ords_prods_group = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_grouping.pkl'))

### Question 4. Data Wrangling the new Customer data set

In [5]:
#Checking columns listed in customers.csv dataframe for unnecessary or illogical names
customers.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [6]:
#Checking the size of the dataframe customers.csv
customers.shape

(206209, 10)

In [7]:
#Checking the basic information of customers.csv
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


### Observations: First Name has only 194950 non-null entries.  This would mean we have 206209-194950 empty cells.  Confirm this below with value.counts function. 

In [8]:
#Double Checking and confirming First Name column for missing values as it is the only one with fewer non-null objects
customers['First Name'].value_counts(dropna = False)

NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: First Name, Length: 208, dtype: int64

### Observations: Confirmed missing 11259 first names, roughly 5.5%, however, it is impossible to determine what their names would be without consulting the source of information.  This shouldn't affect any statistics, so we can leave them blank, at least for now.

In [9]:
#Checking the descriptive statistics of customers.csv
customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [10]:
#Checking the first ten rows of customers.csv 
customers.head(10)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374  
5             2    married   49643  
6             0     single   61746  
7             0     single   63712  
8             1    married  162432  
9             2    married   32072

### Observations: May want to rename Surname to 'Last Name' and make 'STATE' not all caps.  Otherwise, I wouldn't remove any columns just yet

In [11]:
#Rename 'Surname' to Last Name in customers.csv
customers.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [12]:
#Rename 'STATE' to State in customers.csv
customers.rename(columns = {'STATE' : 'State'}, inplace = True)

In [13]:
#Confirm changes to column names
customers.head()

user_id First Name Last Name  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

### Question 5. Complete Data Quality and Consistency Checks

In [14]:
#Check for mixed-types in customers.csv
for col in customers.columns.tolist():
    weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customers[weird]) > 0:
        print(col)

First Name


### Observations: First Name is mixed because it has null values in it.  The rest of the columns are fine.  We can change the data type to String to fix this

In [15]:
#Change the datatype of First Name to a String
customers['First Name'] = customers['First Name'].astype('str')

In [16]:
#Double Checking for Mixed Types
for col in customers.columns.tolist():
    weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customers[weird]) > 0:
        print(col)

In [17]:
#Checking for missing values, though this was already shown above as First Name has missing values.  Confirm after type change
customers.isnull().sum()

user_id         0
First Name      0
Last Name       0
Gender          0
State           0
Age             0
date_joined     0
n_dependants    0
fam_status      0
income          0
dtype: int64

In [18]:
#Find any Duplicate entries in customer.csv and assign them to df_dups
df_dups = customers[customers.duplicated()]

In [19]:
df_dups

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, State, Age, date_joined, n_dependants, fam_status, income]
Index: []

### Observations: Confirmed no duplicates in the dataframe customers.csv

### Question 6. Combining the data with the rest of the prepared Instacart data

In [20]:
#Checking ords_prods_group 
ords_prods_group.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   order_number            int64   
 3   order_day_of_week       int64   
 4   order_hour_of_day       int64   
 5   days_since_prior_order  float64 
 6   product_id              int64   
 7   add_to_cart_order       int64   
 8   reordered               int64   
 9   product_name            object  
 10  aisle_id                int64   
 11  department_id           int64   
 12  prices                  float64 
 13  _merge                  category
 14  price_range_loc         object  
 15  busiest_day             object  
 16  busiest_days            object  
 17  busiest_period_of_day   object  
 18  max_order               int64   
 19  loyalty_flag            object  
 20  average_price           float64 
 21  spendi

In [21]:
#Checking the ords_prods_group df size
ords_prods_group.shape

(32404859, 24)

In [22]:
#Re-checking the customers.csv info
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    206209 non-null  object
 2   Last Name     206209 non-null  object
 3   Gender        206209 non-null  object
 4   State         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [23]:
#Combining the two df using the user_id column
combined = ords_prods_group.merge(customers, on = 'user_id')

In [24]:
#Checking first five rows of combined df
combined.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...        order_freq_flag  First Name  Last Name  Gender  \
0         Soda  ...  Non-Frequent customer       Linda     Nguyen  Female   
1         Soda  ...  Non-Frequent customer       Linda     Nguyen  Female   
2         Soda  ...  Non-Frequent customer       Linda     Nguyen  Female   
3         Soda  ...  Non-Frequent customer       Linda     Nguyen  Female   
4         Soda  ...  Non-Frequent customer       Linda     Nguyen  Female   

     State Age date_joined n_dependants  fam_status income  
0  Alabama  31   2/17/2019            3     married  40423  
1  Alabama  31   2/17/2019            3     married  40423  
2  Alabama  31   2/17/2019            3     married  40423  
3  Alabama  31   2/17/2019            3     married  40423  
4  Alabama  31   2/17/2019            3     married  40423  

[5 rows x 33 columns]

In [25]:
#Checking size of combined df
combined.shape

(32404859, 33)

In [26]:
#Checking basic info of combined df to ensure information is correct
combined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 33 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   order_number            int64   
 3   order_day_of_week       int64   
 4   order_hour_of_day       int64   
 5   days_since_prior_order  float64 
 6   product_id              int64   
 7   add_to_cart_order       int64   
 8   reordered               int64   
 9   product_name            object  
 10  aisle_id                int64   
 11  department_id           int64   
 12  prices                  float64 
 13  _merge                  category
 14  price_range_loc         object  
 15  busiest_day             object  
 16  busiest_days            object  
 17  busiest_period_of_day   object  
 18  max_order               int64   
 19  loyalty_flag            object  
 20  average_price           float64 
 21  spendi

### Question 7. Ensure the notebook contains logical titles, section headings and descriptive code comments

### Question 8 Export this new dataframe as a pickle file so you can continue to use it in Part 2

In [27]:
#Exporting combined customer data set with ords_prods_group to pickle file
combined.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'customer_ords_prods_merged.pkl'))

### Question 9. Save the notebook so that you can send it to your tutor for review